src/E2_economics.py

```diff
@@ -57,138 +64,401 @@
+
+    # Total costs of the assets, capital and operational
+    total_asset_costs_over_lifetime = calculate_total_asset_costs_over_lifetime(
+        dict_asset[COST_INVESTMENT][VALUE], dict_asset[COST_OPERATIONAL_TOTAL][VALUE]
+    )
+    dict_asset.update(
+        {
+            COST_TOTAL: {
+                VALUE: total_asset_costs_over_lifetime,
+                UNIT: economic_data[CURR],
+            }
+        }
+    )
+
+    dict_asset.update(
+        {
+            ANNUITY_TOTAL: {
+                VALUE: dict_asset[COST_TOTAL][VALUE] * economic_data[CRF][VALUE],
+                UNIT: CURR + "/" + UNIT_YEAR,
+            },
+            ANNUITY_OM: {
+                VALUE: dict_asset[COST_OPERATIONAL_TOTAL][VALUE]
+                * economic_data[CRF][VALUE],
+                UNIT: CURR + "/" + UNIT_YEAR,
+            },
+        }
+    )
     return
 
 
-def add_costs_and_total(dict_asset, name, value, total_costs):
-    total_costs += value
-    dict_asset.update({name: {VALUE: value}})
-    return total_costs
+def calculate_total_asset_costs_over_lifetime(
+    costs_investment, cost_operational_expenditures
+):
+    r"""
+    Calculates costs of an asset over whole lifetime
+
+    Parameters
+    ----------
+    costs_investment: float
+        Investment costs over whole lifetime
+
+    cost_operational_expenditures: float
+        Operation and management as well as dispatch expenditures over whole lifetime
+
+    Returns
+    -------
+    total_asset_costs_over_lifetime: float
+        costs of an asset over whole lifetime, including upfront investment costs, development costs, replacement costs, operation and management expenditures, dispatch expenditures
+    """
+    total_asset_costs_over_lifetime = costs_investment + cost_operational_expenditures
+    return total_asset_costs_over_lifetime
+
+
+def calculate_dispatch_expenditures(dispatch_price, flow, asset):
+    r"""
+    Calculated the expenditures connected to an asset due to its dispatch
+
+    Parameters
+    ----------
+    dispatch_price: float, int or pd.Series
+        Dispatch price of an asset (variable costs), ie. how much has to be paid for each unit of dispatch
+        Raises error if type does not match
+        a) lifetime_price_dispatch (taking into account all years of operation)
+        b) price_dispatch (taking into account one year of operation)
+
+    flow: pd.Series
+        Dispatch of the asset
+
+    asset: str
+        Label of the asset
+
+    Returns
+    -------
+    a) Total dispatch expenditures of an asset
+    b) Annual dispatch expenditures of an asset
+    """
+    if isinstance(dispatch_price, float) or isinstance(dispatch_price, int):
+        dispatch_expenditures = dispatch_price * sum(flow)
+    elif isinstance(dispatch_price, pd.Series):
+        dispatch_expenditures = sum(dispatch_price * flow)
+    else:
+        raise UnexpectedValueError(
```

Comment:
I would say it is rather a builtin `TypeError`

Reply:
Changed to TypeError in #525, 44465abb6c1382b61a2cc87a897fbb7de8f1708d